hazelcast/src/main/java/com/hazelcast/sql/impl/client/ReadOptimizedLruCache.java

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import java.util.Comparator;
+import java.util.PriorityQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+/**
+ * Implementation of an LRU cache optimized for read-heavy use cases.
+ * <p>
+ * It stores the entries in a {@link ConcurrentHashMap}, along with the last
+ * access time. It allows the size to grow beyond the capacity, up to
+ * `cleanupThreshold`, at which point the inserting thread will remove a batch
+ * of the eldest items in two passes.
+ * <p>
+ * The cleanup process isn't synchronized to guarantee that the capacity is not
+ * exceeded. The cache is available during the cleanup for reads and writes. If
+ * there's a large number of writes by many threads, the one thread doing the
+ * cleanup might not be quick enough and there's no upper bound on the actual
+ * size of the cache. This is done to optimize the happy path when the keys fit
+ * into the cache.
+ */
+public class ReadOptimizedLruCache<K, V> {
+
+    // package-visible for tests
+    final ConcurrentMap<K, ValueAndTimestamp<V>> cache;
+
+    private final AtomicBoolean cleanupLock = new AtomicBoolean();
+    private final int capacity;
+    private final int cleanupThreshold;
+
+    /**
+     * @param cleanupThreshold The size at which the cache will clean up oldest
+     *     entries in batch. `cleanupThreshold - capacity` entries will be removed.
+     */
+    public ReadOptimizedLruCache(int capacity, int cleanupThreshold) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("capacity <= 0");
+        }
+        if (cleanupThreshold <= capacity) {
+            throw new IllegalArgumentException("cleanupThreshold <= capacity");
+        }
+
+        this.capacity = capacity;
+        this.cleanupThreshold = cleanupThreshold;
+
+        cache = new ConcurrentHashMap<>(cleanupThreshold);
+    }
+
+    public V getOrDefault(K key, V defaultValue) {
+        final V existingValue = get(key);
+        return existingValue != null ? existingValue : defaultValue;
+    }
+
+    public V get(K key) {
+        ValueAndTimestamp<V> valueFromCache = cache.get(key);
+        if (valueFromCache == null) {
+            return null;
+        }
+        valueFromCache.touch();
+        return valueFromCache.value;
+    }
+
+    public void put(K key, V value) {
+        if (value == null) {
+            throw new IllegalArgumentException("Null values are disallowed");
+        }
+
+        ValueAndTimestamp<V> oldValue = cache.put(key, new ValueAndTimestamp<>(value));
+        if (oldValue == null && cache.size() > cleanupThreshold) {
```

Comment:
Isn't it better to trigger cleanup regardless of the key existed or not during a multi thread access? 

```suggestion
        if (cache.size() > cleanupThreshold) {
```